workstreams/Adaptive_Profile_Workstream.md

```diff
@@ -0,0 +1,160 @@
+currently drafted as [Profile
+B](https://github.com/qir-alliance/qir-spec/blob/main/specification/v0.1/7_Profiles.md#profile-a-basic-quantum-functionality)
+on the [specification repository](https://github.com/qir-alliance/qir-spec).
+
+Like the Base Profile, the Adaptive Profile should not rely on any particular
+quantum instruction(s) being available. To target a quantum program to a
+specific backend requires selecting a suitable profile and a compatible quantum
+instruction set. The profile and the quantum instruction set (QIS) selection
+together determine how the program IR should be represented.
+
+A similar specification for the Base Profile can be found
+[here](https://github.com/qir-alliance/qir-spec/blob/main/specification/under_development/profiles/Base_Profile.md).
+The specification of the Adaptive Profile should follow a similar outline for
+the sake of consistency and comprehensiveness of the specification.
+
+## Future Work (out of scope for the Adaptive Profile)
+
+The following features will not be supported in programs that are compliant with
+the Base Profile:
+
+- Composite data types: <br/>
+  The use of composite data types such as [structure
+  types](https://llvm.org/doxygen/group__LLVMCCoreTypeStruct.html) including
+  [tuples](https://github.com/qir-alliance/qir-spec/blob/main/specification/v0.1/1_Data_Types.md#tuples-and-user-defined-types),
+  and [sequential
+  types](https://llvm.org/doxygen/group__LLVMCCoreTypeSequential.html) including
+  [arrays](https://github.com/qir-alliance/qir-spec/blob/main/specification/v0.1/1_Data_Types.md#arrays)
+  is not supported within an Adaptive Profile. This also precludes the use of
+  such data structures for [callable
+  values](https://github.com/qir-alliance/qir-spec/blob/main/specification/v0.1/2_Callables.md);
+  i.e., the usage of subroutines as first class values is not supported within
+  the Base Profile, and the use of [function
+  types](https://llvm.org/doxygen/group__LLVMCCoreTypeFunction.html) is limited
+  to globally declared LLVM functions that may be called as part of program
+  execution.
+
+- Unbounded loops or recursions: <br/>
+  The Adaptive Profile should require that any profile compliant program is
+  known to terminate; it must be possible to statically bind resources and
+  schedule QPU instructions. Supporting direct or indirect function recursions
+  is out of scope for the Adaptive Profile.
+
+- Dynamic qubit allocations and access: <br/>
+  Within the Base Profile, all qubit uses refer directly to a unique qubit id
+  (requires an update of the full specification as part of this workstream).
+  Runtime functions for qubit allocation/release are not available, and the lack
+  of support for local variables and composite data types prevents any qubit
+  aliasing. Prohibiting classical computations and branching based on
+  measurement results furthermore precludes any dynamic qubit access.
```

Comment:
Seems like a limiting requirement. Should getting concrete on what is/isn't allowed here be deferred to the actual workstream to decide, rather than stated in this outline for the workstream?

Reply:
It is limiting, yes, though only a limitation for what this profile/workstream covers. It helps to have some very rough "in scope/out of scope" guidelines to set expectations on what the profile is for. Dynamic memory management is a whole different level of complexity to support, and hence out of scope for this workstream. Other things that are more feasible to support are listed in open questions and may be either in- or out of scope depending on what the working group deems reasonable. 

Maybe I should mention that language frontends do not directly compile into a profile (QAT for example has a pass to translate allocations/deallocations via rt functions into static qubit ids). Furthermore, this does largely not impact what can be in the source language (the limitation here for the source language is that the memory, like the execution time, needs to be known to be finite and there may be overhead in the resource allocation depending on how smart the compiler is).